### DISCUSSION TITLE
ESP32 build error

### DISCUSSION THREAD

**Discussion by mdaeron** [micropython/micropython#12309] [ESP32] [answered]
I'm trying to compile firmware for an ESP32 board using the following commands:

```sh
DIR=`pwd`

git clone --depth 1 -b v5.0.2 --recursive https://github.com/espressif/esp-idf.git
cd $DIR/esp-idf
git submodule update --init --recursive
./install.sh
source export.sh

git clone https://github.com/glenn20/micropython.git --depth 1 --branch espnow-g20-v1.20.0 --single-branch $DIR/micropython

cd $DIR/micropython
make -C mpy-cross

cd $DIR/micropython/ports/esp32
make submodules
make
```

I'm getting the following error in the final `make` call:

```
-- Configuring incomplete, errors occurred!
HINT: The component esp_adc_cal could not be found. This could be because: component name was misspelled, the component was not added to the build, the component has been moved to the IDF component manager or has been removed and refactored into some other component.
Please look out for component in 'https://components.espressif.com' and add using 'idf.py add-dependency' command.
Refer to the migration guide for more details about moved components.
Refer to the build-system guide for more details about how components are found and included in the build.
cmake failed with exit code 1, output of the command is in the /Users/daeron/ShareBox/uPy/firmware/esp32-s2-mini-low-energy-boot/micropython/ports/esp32/build-GENERIC/log/idf_py_stderr_output_11698 and /Users/daeron/ShareBox/uPy/firmware/esp32-s2-mini-low-energy-boot/micropython/ports/esp32/build-GENERIC/log/idf_py_stdout_output_11698
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [all] Error 1
```

I'm on MacOS 12.3.1.

Is there something obviously wrong in how I do things?

**Comment by andrewleech** [comment, marked as answer]
It looks like you're trying to compile an older branch on @glenn20 's fork of micropython for espnow support?

This older v1.20 branch presumably doesn't work with idf 5.0 as that only very recently had support added in the main micropython repo.

I'm pretty sure the espnow support has been merged here on the main repo, are you sure you need this forked branch?

**Comment by glenn20** [comment]
Yes, as @andrewleech says, that branch represents my espnow patches applied against the micropython v1.20.0 codebase. Micropython v1.20.0 does not support idf 5.0.  The issue is that you are using the post-idf5.0 build instructions (from main branch) for a v1.20.0 codebase.

You could either:

- Build an espnow-enabled micropython v1.20.0 (with idf 4.2 or v4.4):
  - Follow the build instructions for Micropython v1.20.0 (see `ports/esp32/README.md` you downloaded from my branch or [README.md](https://github.com/micropython/micropython/blob/v1.20.0/ports/esp32/README.md).
- Build from the Micropython master branch (which now includes espnow) rather than my branch:
  - Follow the post idf5.0 build instructions in Micropython main branch [README.md](https://github.com/micropython/micropython/blob/master/ports/esp32/README.md).
- Just download a nightly build image which now includes espnow: https://micropython.org/download/?port=esp32

**Comment by mdaeron** [comment]
Thanks to both of you, that makes sense. I got this to work:

```sh
#!/bin/zsh

DIR=`pwd`
BOARD=LOLIN_S2_MINI

# COMMENT NEXT FOUR LINES IF ESP-IDF IS ALREADY INSTALLED
git clone --depth 1 -b v5.0.2 --recursive https://github.com/espressif/esp-idf.git
cd $DIR/esp-idf
git submodule update --init --recursive
./install.sh

cd $DIR/esp-idf
source export.sh

git clone https://github.com/micropython/micropython.git --depth 1 --branch master --single-branch $DIR/micropython

# THE NEXT THREE LINES ARE FOR MINIMIZING STARTUP TIME (CF https://github.com/glenn20/upy-esp32-experiments)
# cat $DIR/main.py >> $DIR/micropython/ports/esp32/modules/_boot.py
# echo "\nCONFIG_BOOTLOADER_SKIP_VALIDATE_IN_DEEP_SLEEP=y" >> $DIR/micropython/ports/esp32/boards/sdkconfig.base
# gsed -i "/void boardctrl_startup(void) {/a\    if (esp_reset_reason() == ESP_RST_DEEPSLEEP) {\n         return;\n     }" $DIR/micropython/ports/esp32/main.c

cd $DIR/micropython
make -C mpy-cross

cd $DIR/micropython/ports/esp32
make submodules BOARD=$BOARD
make BOARD=$BOARD
make erase BOARD=$BOARD PORT=`ls /dev/tty.usb*`
make deploy BOARD=$BOARD PORT=`ls /dev/tty.usb*`
```

As you can see @glenn20 I'm playing with your optimizations for low-power startup of ESPNow devices. So far, using the script above, I got my board to blink a LED according to code in `ports/esp32/modules/_boot.py`, so that's encouraging.

Quick additional question: I assume that with these startup optimizations, I lose access to `rshell` and the REPL, right?

Thanks again!

**Reply by glenn20** [reply]
> As you can see @glenn20 I'm playing with your optimizations for low-power startup of ESPNow devices. So far, using the script above, I got my board to blink a LED according to code in ports/esp32/modules/_boot.py, so that's encouraging.

Cool - I hope it all goes well.

> Quick additional question: I assume that with these startup optimizations, I lose access to rshell and the REPL, right?

If the app in `_boot.py` does not exit or goes back into deepsleep, then the repl will not be started. However, if execution of `_boot.py` ends and does NOT go back into deepsleep and mounts the flash, then the device will proceed to boot up normally - executing `boot.py`, `main.py` and starting the repl. You could define some conditions so that the device proceeds to boot normally to the repl (which is useful for test, dev and debugging).

For my test devices they were setup to only go back into deepsleep if waking from deepsleep. This means that on a normal power up or hard reset they would boot up normally. If I call `machine.deepsleep()` from there, it kicks off the wake from deepsleep and back to deepsleep from `_boot.py` cycle.

NOTE: if you have skipped the call to `nvs_flash_init()` in `boardctrl_startup()` you will not be able to start wifi.

**Comment by mdaeron** [comment]
> For my test devices they were setup to only go back into deepsleep if waking from deepsleep. This means that on a normal power up or hard reset they would boot up normally. If I call `machine.deepsleep()` from there, it kicks off the wake from deepsleep and back to deepsleep from `_boot.py` cycle.

Oh, that's neat, thanks for the tip. Do I need to do anything in `_boot.py` to explicitly mount the flash?

**Reply by glenn20** [reply]
> Oh, that's neat, thanks for the tip. Do I need to do anything in `_boot.py` to explicitly mount the flash?

Yes. Take a look at `ports/esp32/modules/_boot.py`. In my tests, I just added a line `import  _preboot` at the start of `_boot.py` and added my app in `ports/esp32/modules/_preboot.py`. You can see that in the files in the branch I used for those tests.

It is recommended to do all this with a custom manifest instead of placing your files to be frozen into the modules dir, but I wanted to keep my test changes contained within the microphone folder.

**Reply by mdaeron** [reply]
Cool, thanks.

**Comment by mdaeron** [comment]
Thanks!
